Loki: implement decolorize filter

vmax opened the pull request:
**What this PR does / why we need it**:
Implement `decolorize` filter that get rids of ANSI color codes in the log message.

**Which issue(s) this PR fixes**:
Fixes #7601

**Special notes for your reviewer**:

It's my first time writing Golang, so please let me know if I missed anything obvious 🙂 

**Checklist**
- [x] Reviewed the `CONTRIBUTING.md` guide
- [x] Documentation added
- [x] Tests updated
- [x] `CHANGELOG.md` updated
- [ ] Changes that require user attention or interaction to upgrade are documented in `docs/sources/upgrading/_index.md`

dannykopping reviewed (commented):
I love this idea @vmax! Can you please include some unit tests?

dannykopping commented:
We can do this in a separate PR, but it would also be great to have this as a promtail stage

vmax commented:
@dannykopping I've implemented a unit test for `Decolorizer` in https://github.com/grafana/loki/pull/7602/commits/c748326259bc921acb0015a414c73f4ed1923150; modified tests for [parser](https://github.com/grafana/loki/pull/7602/commits/a9d4298a100bf9e643f042b171a1901eb75d47b0) and for [lexer](https://github.com/grafana/loki/pull/7602/commits/55bbf2929f32bf5eec84366bd89142ff12f88406)

vmax commented:
@dannykopping I've also added `decolorize` [pipeline](https://github.com/grafana/loki/pull/7602/commits/f0be8791832ee6aa740881be6647a26fb0a1dde2) and [tests for it](https://github.com/grafana/loki/pull/7602/commits/c309f2d1e547166d841aeb530d5df41cf4524356)

vmax commented:
@dannykopping I've amended the commit to include:
- tests for `Decolorizer`
- updated tests for parser and lexer
- `decolorize` stage for Promtail and tests for it

vlad-diachenko commented on CHANGELOG.md:
```diff
@@ -24,6 +24,7 @@
 * [7264](https://github.com/grafana/loki/pull/7264) **bboreham**: Chunks: decode varints directly from byte buffer, for speed.
 * [7263](https://github.com/grafana/loki/pull/7263) **bboreham**: Dependencies: klauspost/compress package to v1.15.11; improves performance.
 * [7270](https://github.com/grafana/loki/pull/7270) **wilfriedroset**: Add support for `username` to redis cache configuration.
+* [7602](https://github.com/grafana/loki/pull/7602) **vmax**: Add decolorize filter to easily parse colored logs.
```
could you please move your changelog entries to new sections from https://github.com/grafana/loki/pull/7606

you added your entries to Loki 2.7.0 changelog.

vmax replied:
@vlad-diachenko rebased & moved to `Unreleased` section in https://github.com/grafana/loki/pull/7602/commits/bb901e4691a015976ab37aa33bdb7adad41cac7d

vlad-diachenko reviewed (changes requested):
Looks great. thanks a lot for your contribution !!!
it would be awesome if you could also add a section to `docs/sources/logql/log_queries.md` to add documentation on how it works. 
after the merge, the doc will be automatically published to Loki docs page here https://grafana.com/docs/loki/next/logql/log_queries/

vmax commented on pkg/logql/syntax/lex.go:
```diff
@@ -66,8 +66,9 @@ var tokens = map[string]int{
-	OpFmtLine:  LINE_FMT,
+	OpFmtLabel:   LABEL_FMT,
+	OpFmtLine:    LINE_FMT,
+	OpDecolorize: DECOLORIZE,
```
@vlad-diachenko indeed! addressed in https://github.com/grafana/loki/pull/7602/commits/f5dabc955ff1ef856d55c35719f808fe3d6bc139

vlad-diachenko reviewed (approved):
Looks brilliant 💎
